File: app/src/main/java/com/example/mypc/fastfoodfinder/ui/store/StoreInfoDialogFragment.java
package com.example.mypc.fastfoodfinder.ui.store;

import android.app.Dialog;
import android.graphics.Point;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.view.Display;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.example.mypc.fastfoodfinder.R;
import com.example.mypc.fastfoodfinder.activity.StoreDetailActivity;
import com.example.mypc.fastfoodfinder.model.Store.Store;
import com.example.mypc.fastfoodfinder.utils.DisplayUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by taq on 26/11/2016.
 */

public class StoreInfoDialogFragment extends DialogFragment {

    public static StoreInfoDialogFragment newInstance(Store store) {
        Bundle args = new Bundle();
        args.putParcelable(StoreDetailActivity.STORE, store);
        StoreInfoDialogFragment fragment = new StoreInfoDialogFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @BindView(R.id.store_name)
    TextView tvStoreName;
    @BindView(R.id.view_detail)
    TextView tvViewDetail;
    @BindView(R.id.store_address)
    TextView tvStoreAddress;
    @BindView(R.id.call_direction)
    View vCallDirection;
    @BindView(R.id.save_this)
    Button btnAddToFavorite;

    private StoreDialogActionListener mListener;
    public CallDirectionViewHolder cdvh;

    // tên chuối thiệt
    public interface StoreDialogActionListener {
        void onDirection(Store store);
        void onAddToFavorite(int storeId);
    }

    public StoreInfoDialogFragment() {

    }

    public void setDialogListen(StoreDialogActionListener listener)
    {
        mListener = listener;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_store_info, container);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);
        cdvh = new CallDirectionViewHolder(vCallDirection);

        final Store store =  getArguments().getParcelable(StoreDetailActivity.STORE);

        tvStoreName.setText(store.getTitle());
        tvStoreAddress.setText(store.getAddress());
        tvViewDetail.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getActivity().startActivity(StoreDetailActivity.getIntent(getContext(), store));
            }
        });

        cdvh.btnCall.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (store.getTel() != null && !store.getTel().equals("")) {
                    startActivity(DisplayUtils.getCallIntent(store.getTel()));
                } else {
                    Toast.makeText(getActivity(), "The ic_store doesn't have number phone!", Toast.LENGTH_SHORT).show();
                }
            }
        });
        cdvh.btnDirection.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mListener.onDirection(store);
            }
        });

        btnAddToFavorite.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mListener.onAddToFavorite(store.getId());
                dismiss();
            }
        });
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);
        dialog.getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        return dialog;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getDialog().getWindow().getAttributes().windowAnimations = R.style.DialogAnimation;
    }

    @Override
    public void onResume() {
        Window window = getDialog().getWindow();
        Point size = new Point();
        Display display = window.getWindowManager().getDefaultDisplay();
        display.getSize(size);
        window.setLayout((int)(0.8 * size.x), WindowManager.LayoutParams.WRAP_CONTENT);
        window.setGravity(Gravity.CENTER);
        super.onResume();
    }
}

File: app/src/main/kotlin/com/iceteaviet/fastfoodfinder/ui/store/StoreInfoDialogFragment.kt
package com.iceteaviet.fastfoodfinder.ui.store

import android.app.Dialog
import android.content.pm.PackageManager
import android.graphics.Point
import android.os.Bundle
import android.view.*
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.fragment.app.DialogFragment
import com.iceteaviet.fastfoodfinder.R
import com.iceteaviet.fastfoodfinder.data.remote.store.model.Store
import com.iceteaviet.fastfoodfinder.ui.store.StoreDetailActivity.Companion.KEY_STORE
import com.iceteaviet.fastfoodfinder.utils.*
import kotlinx.android.synthetic.main.fragment_store_info.*

/**
 * Created by taq on 26/11/2016.
 */

class StoreInfoDialogFragment : DialogFragment() {

    lateinit var cdvh: StoreDetailAdapter.CallDirectionViewHolder
    lateinit var tvStoreName: TextView
    lateinit var tvViewDetail: TextView
    lateinit var tvStoreAddress: TextView
    lateinit var vCallDirection: View
    lateinit var btnAddToFavorite: Button

    private var store: Store? = null

    private var mListener: StoreDialogActionListener? = null

    fun setDialogListen(listener: StoreDialogActionListener) {
        mListener = listener
    }

    @Nullable
    override fun onCreateView(@NonNull inflater: LayoutInflater, @Nullable container: ViewGroup?, @Nullable savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.fragment_store_info, container)
    }

    override fun onViewCreated(@NonNull view: View, @Nullable savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        tvStoreName = store_name
        tvViewDetail = view_detail
        tvStoreAddress = store_address
        vCallDirection = call_direction
        btnAddToFavorite = btn_fav

        cdvh = StoreDetailAdapter.CallDirectionViewHolder(vCallDirection) // TODO: Check this !!

        if (arguments != null)
            arguments?.let { store = it.getParcelable(KEY_STORE) }

        if (store == null)
            return

        tvStoreName.text = store!!.title
        tvStoreAddress.text = store!!.address
        tvViewDetail.setOnClickListener {
            store?.let { activity?.startActivity(StoreDetailActivity.getIntent(context!!, it)) }
        }

        cdvh.btnCall.setOnClickListener {
            if (!isEmpty(store!!.tel)) {
                if (isCallPhonePermissionGranted(context!!))
                    startActivity(getCallIntent(store!!.tel!!))
                else
                    requestCallPhonePermission(this@StoreInfoDialogFragment)
            } else {
                Toast.makeText(activity, R.string.store_no_phone_numb, Toast.LENGTH_SHORT).show()
            }
        }
        cdvh.btnDirection.setOnClickListener { mListener!!.onDirection(store) }

        btnAddToFavorite.setOnClickListener {
            mListener!!.onAddToFavorite(store!!.id)
            dismiss()
        }
    }

    @NonNull
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        dialog.window!!.requestFeature(Window.FEATURE_NO_TITLE)
        return dialog
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        dialog.window!!.attributes.windowAnimations = R.style.DialogAnimation
    }

    override fun onResume() {
        val window = dialog.window
        val size = Point()
        val display = window!!.windowManager.defaultDisplay
        display.getSize(size)
        window.setLayout((0.8 * size.x).toInt(), WindowManager.LayoutParams.WRAP_CONTENT)
        window.setGravity(Gravity.CENTER)
        super.onResume()
    }

    override fun onRequestPermissionsResult(requestCode: Int, @NonNull permissions: Array<String>, @NonNull grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        when (requestCode) {
            REQUEST_CALL_PHONE -> {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    startActivity(getCallIntent(store!!.tel!!))
                } else {
                    Toast.makeText(activity, R.string.permission_denied, Toast.LENGTH_SHORT).show()
                }
            }

            else -> {
            }
        }
    }

    // tên chuối thiệt
    interface StoreDialogActionListener {
        fun onDirection(store: Store?)

        fun onAddToFavorite(storeId: Int)
    }

    companion object {
        fun newInstance(store: Store): StoreInfoDialogFragment {
            val args = Bundle()
            args.putParcelable(KEY_STORE, store)
            val fragment = StoreInfoDialogFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
